### PR TITLE
[Feature]: support FP8 communication in DDP, FSDP, Gemini

### LINKED ISSUE #5926: [Feature]: support FP8 communication in DDP
Request FP8 compressed communication for gradient communication (all-reduce) in DDP.

### LINKED ISSUE #5927: [Feature]: support FP8 communication in FSDP
Request FP8 compressed communication for gradient communication (scatter-reduce) in FSDP, and parameter  communication (all-gather) in FSDP.

### LINKED ISSUE #5943: [Feature]: support FP8 communication in Gemini
Request FP8 compressed communication for parameter all-gather in Gemini.

### PR DESCRIPTION
**Pull request by BurkeHulk**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

resolved #5926, resolved #5927, resolved #5943

## 📝 What does this PR do?

Implement FP8 compressed communication hook for gradient all-reduce in torch DDP, 
gradient scatter-reduce in FSDP, and parameter all-gather in FSDP. 

For training accuracy validation, please refer to the results below.

| **DDP**  | **BERT - 4 GPUs** | **GPT2 - 4 GPUs** |
|----------|-------------------|-------------------|
| FP32     |             0.832 |            0.7089 |
| FP8-e4m3 |             0.829 |            0.7077 |
| FP8-e5m2 |              0.820 |            0.7089 |

| **FSDP-grad** | **BERT - 4 GPUs** | **BERT - 8 GPUs** |
|---------------|-------------------|-------------------|
| FP16          |            0.8344 |            0.8385 |
| FP8-e4m3      |            0.8304 |            0.8408 |
| FP8-e5m2      |            0.8344 |            0.8304 |

| **FSDP-params** | **BERT - 4 GPUs** | **BERT - 8 GPUs** |
|-----------------|-------------------|-------------------|
| FP16            |            0.8344 |            0.8385 |
| FP8-e4m3        |            0.8298 |            0.8443 |
| FP8-e5m2        |            0.8281 |            0.8362 |

| **FSDP-grad+params** | **BERT - 4 GPUs** |
|----------------------|-------------------|
| FP16                 |            0.8344 |
| FP8-e4m3             |            0.8368 |
| FP8-e5m2             |            0.8373 |

| **Gemini** | **BERT - 4 GPUs** |
|------------|-------------------|
| FP16       |             0.8350 |
| FP8-e4m3   |            0.8373 |
| FP8-e5m2   |            0.8362 |

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/quantization/fp8.py [hunk 1]
```diff
@@ -170,3 +173,229 @@ def cast_from_fp8_pipeline(inp: Any, del_metadata=True) -> None:
...
+        output_chunks.append(torch.empty_like(ret))
+        output_scale_list.append(torch.empty_like(scale))
+    dist.all_to_all(output_chunks, cast_input_list, group=group)
+    dist.all_to_all(output_scale_list, scale_list, group=group)
+
+    summed_out = torch.zeros_like(output_chunks[0]).to(input_type)
+    for scale, out in zip(output_scale_list, output_chunks):
+        out = out.view(fp8_type)
+        summed_out += cast_from_fp8(out, scale, input_type)
+    output.data = summed_out
```
Comment (ver217) [inline colossalai/quantization/fp8.py:212]: Should it divide by world size?

Comment (BurkeHulk) [reply colossalai/quantization/fp8.py:212]: @GuangyaoZhang , do you need division here for shardformer?

Comment (GuangyaoZhang) [reply colossalai/quantization/fp8.py:212]: No. Only sum op is needed for reduce scatter in shardformer.

**Review by ver217 — APPROVED** [review]

**Review by ver217 — APPROVED** [review]
